File: app/src/main/java/net/squanchy/schedule/view/ScheduleViewPagerAdapter.java
package net.squanchy.schedule.view;

import android.content.Context;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.novoda.viewpageradapter.ViewPagerAdapter;

import java.util.Collections;
import java.util.List;

import net.squanchy.R;
import net.squanchy.schedule.domain.view.Event;
import net.squanchy.schedule.domain.view.SchedulePage;

import org.joda.time.LocalDateTime;

public class ScheduleViewPagerAdapter extends ViewPagerAdapter<ScheduleDayPageView> {

    private static final String TITLE_FORMAT_TEMPLATE = "EEE d";

    private final Context context;

    private List<SchedulePage> pages = Collections.emptyList();

    @Nullable
    private OnEventClickedListener listener;

    public ScheduleViewPagerAdapter(Context context) {
        this.context = context;
    }

    public void updateWith(List<SchedulePage> pages, OnEventClickedListener listener) {
        this.pages = pages;
        this.listener = listener;
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return pages.size();
    }

    @Override
    protected ScheduleDayPageView createView(ViewGroup container, int position) {
        return (ScheduleDayPageView) LayoutInflater.from(context)
                .inflate(R.layout.view_page_schedule_day, container, false);
    }

    @Override
    protected void bindView(ScheduleDayPageView view, int position) {
        List<Event> events = pages.get(position).events();
        view.updateWith(events, listener);
    }

    @Override
    public CharSequence getPageTitle(int position) {
        LocalDateTime date = pages.get(position).date();
        return date.toString(TITLE_FORMAT_TEMPLATE).toUpperCase();
    }

    public String getPageDayId(int position) {
        return pages.get(position).dayId();
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

    public interface OnEventClickedListener {

        void onEventClicked(Event event);
    }
}

File: app/src/main/java/net/squanchy/schedule/view/ScheduleViewPagerAdapter.kt
package net.squanchy.schedule.view

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.novoda.viewpageradapter.ViewPagerAdapter
import net.squanchy.R
import net.squanchy.schedule.domain.view.Event
import net.squanchy.schedule.domain.view.SchedulePage
import java.util.Locale

class ScheduleViewPagerAdapter(context: Context) : ViewPagerAdapter<ScheduleDayPageView>() {

    private lateinit var listener: (Event) -> Unit

    private var pages = emptyList<SchedulePage>()

    private val inflater = LayoutInflater.from(context)
    private val viewPool = RecyclerView.RecycledViewPool()
    private var showRoom: Boolean = false
    private var showFavorites: Boolean = false

    fun updateWith(pages: List<SchedulePage>, showRoom: Boolean, showFavorites: Boolean, listener: (Event) -> Unit) {
        this.pages = pages
        this.showRoom = showRoom
        this.showFavorites = showFavorites
        this.listener = listener
        notifyDataSetChanged()
    }

    override fun getCount() = pages.size

    override fun createView(container: ViewGroup, position: Int): ScheduleDayPageView {
        val recyclerView = inflater.inflate(R.layout.view_page_schedule_day, container, false) as ScheduleDayPageView
        recyclerView.setRecycledViewPool(viewPool)
        return recyclerView
    }

    override fun bindView(view: ScheduleDayPageView, position: Int) {
        val events = pages[position].events
        view.updateWith(events, showRoom, showFavorites, listener)
    }

    override fun getPageTitle(position: Int): CharSequence? {
        val date = pages[position].date
        return date.toString(TITLE_FORMAT_TEMPLATE).toUpperCase(Locale.getDefault())
    }

    override fun isViewFromObject(view: View, anObject: Any) = view === anObject

    companion object {
        private const val TITLE_FORMAT_TEMPLATE = "EEE d"
    }
}
